Fix Skillchains sometimes triggering before the 3 second window

DiscipleOfEris opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

The code was not checking for the delay if it was a multi skillchain. That is, the 3rd or later WS in the skillchain.

Fixes #2533

## Steps to test these changes

See #2533

DiscipleOfEris committed:
Fix Skillchains sometimes triggering before the 3 second window

The code was not checking for the delay if it was a multi skillchain. That is, the 3rd or later WS in the skillchain.

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -3729,28 +3729,29 @@ namespace battleutils
             }
             // Previous effect exists
-            else if (PSCEffect && PSCEffect->GetTier() == 0)
+            else if (PSCEffect->GetStartTime() + 3s < server_clock::now())
```
Can SCEffect be nullptr here?

TeoTwawki replied:
there's something else to consider here. we _perceive_ a 3 second delay on retail, but there's lag time between the player and Japan. Actually, is it 3 seconds? I've not timed it. I'm unsure if BG accounts for this but gives a very vague timeframe on its skillchains page near the top that def sounds like nobody actually documented real world tests. Potentially animation times of the involved weaponskills matter. I'd have never thought something so fundamental to the game would make me say "I don't know" but here I am, doubting myself.

zach2good replied:
Sounds like a good candidate for a setting, but also, 2.4s is very suspiciously nearby

DiscipleOfEris replied:
This is in an else of an else branch. To reach the first else branch, at least one of `PSCEffect` and `PCBEffect` must not be `nullptr`. To reach the second branch, `PCBEffect` must be `nullptr`. As such, to reach this branch `PSCEffect` cannot be `nullptr`. That said, the code might look nicer with a `nullptr` check anyway.

TeoTwawki replied:
small clarification: it was 3 before the contributor got here, so I don't mean we need to check it before merging this. they are just making the timing it already said actually work when it wasn't. but we should in the future confirm the timing was even right.

zach2good replied:
I'm on my phone, I can see the upper check now, thanks!

DiscipleOfEris replied:
The SC delay should be relatively easy to test (for someone with a retail account), including if it's affected by the weaponskill and/or skillchain duration.

WinterSolstice8 replied:
you can't trust the unmodified client for the timing based on the log message for skillchains. some of them (like Transfixion/fragmentation) emit the message early, and others a bit later than most and you can screw up the skillchain timing. Better testing would need an addon to see exactly when the packet came through

this will get checked in the Great Rewrite:tm:

TeoTwawki replied:
> Sounds like a good candidate for a setting, but also, 2.4s is very suspiciously nearby

as it happens time to japan is about 300ms. 300 there, 300 back. 600ms +2.4 seconds = 3 seconds!

DiscipleOfEris replied:
@WinterSolstice8 Well yeah. I had exactly that in mind, since I do a lot of packet sniffing. That said, I suppose my definition of "easy to test" isn't exactly standard, so point.

DiscipleOfEris replied:
I'd basically just use a script to control the input timings of the WSs involved, choosing some that have a low animation time. Then adjust the wait period down until the SC triggers about half the time.

So like, `//input /ws "Shining Strike" <t>; wait 2.9; send myalt input /ws "Gust Slash" <t>;`. Where I keep adjusting that 2.9 to find the sweet spot. Then switch to a different set of WSs to see if the window differs.

And perform the same test but for double SCs (using a SAM or three alts or something), to see if the animation duration of the SC affects the timing window of the 3rd WS.

DiscipleOfEris committed:
Cleanup nullptr check

This `nullptr` check is technically unnecessary, but it looks cleaner with it in.

DiscipleOfEris committed:
Formatting compliance

Failed a Clang check. I think it's this double empty line.

DiscipleOfEris commented:
The sanity check failed again... Anyone have a better approach to running these sanity checks locally before committing?

TeoTwawki commented:
> The CPP Formatting sanity check failed again... Anyone have a better approach to running these sanity checks locally before committing?

I have legit never tried to do that ever. But your stuff looks fine and I am gonna say the checker here is wrong not you : its griping about some removed whitespace, and I _want_ whitespace to be removed so the checker can stuff it :stuck_out_tongue:

claywar commented:
That's what the error is showing.  CI puts out a diff for what was there, and what it expects it to be replaced with.  It's not saying add whitespace.

DiscipleOfEris committed:
CPP Formatting sanity check

Removing spaces from the empty line.

TeoTwawki commented:
> That's what the error is showing. CI puts out a diff for what was there, and what it expects it to be replaced with. It's not saying add whitespace.

Then the display is backwards from what is expected. The diff shows whitespace being removed, and we _want_ whitespace removed. When I look at CI things my expectation is to be shown "where's where you are wrong". Not "here is what to do", Thats just fuckin confusing.

TeoTwawki commented:
incidentally, just always trimming trailing when you save is a good practice to be in anyway

DiscipleOfEris commented:
I guess I understand how to read the sanity check now.

TeoTwawki commented:
> I guess I understand how to read the sanity check now.

doin better than me then, apparently! _mutter grumble_
